tbot_contrib/locking.py

```diff
@@ -0,0 +1,166 @@
+import contextlib
+import typing
+import tbot
+import time
+from tbot.machine import machine, linux
+
+
+class MachineLock(machine.PreConnectInitializer):
+    """
+    Machine locking implementation based on Python and bash.
+    """
+    @property
+    def lock_name(self) -> str:
+        return self.__class__.__name__
+
+    @property
+    def lock_expiry(self) -> int:
+        return -1
+
+    @property
+    def lock_pool_max(self) -> int:
+        return 1
+
+    @property
+    def lock_sleep(self) -> int:
+        return 1
+
+    lock_no = -1
+    lock_file = ""
+
+    @contextlib.contextmanager 
+    def _init_pre_connect(self) -> typing.Iterator:
+        with tbot.ctx.request(tbot.role.LabHost) as lh:
+            result = lh.exec("mktemp", "lock.XXX")
+            if result[0] == 0:
+                tempfile = result[1].strip()
+                if self.lock_expiry < 1:
+#                    lh.exec("echo", "0", linux.RedirStdout(tempfile))
+                    lh.exec("echo", "0", linux.Raw(">"), tempfile)
+                else: # Write to file the Unix time timestamp of expiry
+#                    lh.exec("echo", str(int(time.time())+self.lock_expiry), linux.RedirStdout(tempfile))
+                    lh.exec("echo", str(int(time.time())+self.lock_expiry), linux.Raw(">"), tempfile)
+            else:
+                raise Exception("Could not create tempfile")
+            repeat = True
+            while (repeat == True):
+                repeat = False
+                lock_no = -1
+                for count in range(self.lock_pool_max): # Iterate through locks
+                    self.lock_file = f"{self.lock_name}{count}"
+                    print(f"Try locking on {count} using {self.lock_file}...")
+                    result = lh.exec("ln", tempfile, self.lock_file) # Is atomic 
+                    if result[0] == 0: # We got a lock, we're done
```

Comment:
Use `LinuxMachine.test()` if you just care about success of a command:

```suggestion
                    if lh.test("ln", tempfile, self.lock_file): # We got a lock, we're done
```